### PR TITLE
Graphical TFT fixes, cleanup

### PR DESCRIPTION
**Pull request by tpruvot**
lcd_put_wchar_max needed for 480x320 too

followup of #20838

### CODE REVIEW

**Comment by tpruvot** [comment]
on hold, need to do more tests... Level Corners output seems messed up in color ui... but its not related to this PR changes

image

**Comment by tpruvot** [comment]
to improve and test on 480x320, but with the last commit ive that 
image

**Comment by tpruvot** [comment]
k, its the best i can do for now....

image

**Comment by tpruvot** [comment]
@rhapsodyv i let you try on the 480x320 ?

**Comment by rhapsodyv** [comment]
@tpruvot Idea: instead of having a cursor to drawing, it would not be better to lcd_put_* just concatenate data to the current tft_string, and then we just add a new method to be called: tft.draw_current_string.

With this approach, we just need add one call at the final of all lcd_put_* calls, instead of having to add customized lcd_move_to everywhere.

What do you think?

**Comment by tpruvot** [comment]
I rebased + squashed the 2 last commits in one...

There may be other screens to check i'm unable to test for now :

Marlin/src/feature/power_monitor.cpp:60 (amps/volts/watts)
Marlin/src/lcd/menu/menu_mixer.cpp:60
Marlin/src/lcd/menu/menu_mixer.cpp:117
Marlin/src/lcd/menu/menu_tune.cpp:75 (babystep, testable if i enable it)

will try also to print a matrix of all lines/columns, ive the feeling the rows are truncated...

also lcd_put_int(int) is not implemented in color ui

**Comment by tpruvot** [comment]
tested ok, but i still get crash/weird issue on a matrix test

```
#if ENABLED(TFT_COLOR_UI)
    for (y=0; y<7; y++) {
      for (uint8_t x=0; x<16; x++) {
        lcd_moveto(x, y);
        lcd_put_int(x);
      }
    }
#endif
```
with x<14 that displays it

**Comment by tpruvot** [comment]
k, my test issue was stack problem when doing too much lcd_put_int(x) on the same row...
for (uint8_t x=5; x<17; x++) works

x out of range tested ok & ignored so... 7x20 on 320x240

File: Marlin/src/lcd/tft/ui_320x240.cpp [hunk 1]
```diff
@@ -682,12 +675,16 @@ void lcd_gotopixel(const uint16_t x, const lcd_uint_t y) {
   tft.set_background(COLOR_BACKGROUND);
 }
 
+void lcd_moveto(const lcd_uint_t col, const lcd_uint_t row) {
+  lcd_gotopixel(col * (TFT_COL_WIDTH), row * (MENU_ITEM_HEIGHT));
```
Comment (tpruvot) [inline Marlin/src/lcd/tft/ui_320x240.cpp]: i fear an overflow, i think lcd_uint_t is uint8_t for the 320x240 (320 > 255)
so >> (TFT_COL_WIDTH) * col  ?

Comment (tpruvot) [reply Marlin/src/lcd/tft/ui_320x240.cpp]: hmm seems ok in fact... c++ vs c i guess

**Comment by tpruvot** [comment]
before merging, i think we can reenable Marlin/src/inc/SanityCheck.h:806
```diff
-  #elif ENABLED(BABYSTEP_DISPLAY_TOTAL) && ANY(TFT_320x240, TFT_320x240_SPI, TFT_480x320, TFT_480x320_SPI)
-    #error "New Color UI (TFT_320x240, TFT_320x240_SPI, TFT_480x320, TFT_480x320_SPI) does not support BABYSTEP_DISPLAY_TOTAL yet."
```
code is similar, in Marlin/src/lcd/menu/menu_tune.cpp

at least its built now... but didnt test in real conditions

File: Marlin/src/lcd/tft/ui_320x240.cpp (hunk 1 above)
Comment (thinkyhead) [reply Marlin/src/lcd/tft/ui_320x240.cpp]: I may be mistaken, but I believe multiplication by a constant integer promotes to a `signed int` before taking the final result.

**Comment by thinkyhead** [comment]
> reenable … SanityCheck

The sanity check is already enabled.
Do you mean, we should remove it?

**Comment by tpruvot** [comment]
yes, display is not perfect but its readable, erk was not easy to find it :p but Z Probe Babystep is ok

image

need to start a print to access the tune menu... painful

"last row" is not the good place i think... that overflow the buttons... and you increased the line height to 34... imo should be 30 or 32 (on 320x240)

erk, code using lcd_put_u8str_P(GET_TEXT(MSG_BABYSTEP_TOTAL)); which doesnt increment our X cursor

**Comment by tpruvot** [comment]
need 32, if we use the +2

right now with your common ui change, its doing that : 

image

**Comment by tpruvot** [comment]
Marlin/src/lcd/menu/menu_tune.cpp

```cpp
    if (ui.should_draw()) {
      const float spm = planner.steps_to_mm[axis];
      MenuEditItemBase::draw_edit_screen(msg, BABYSTEP_TO_STR(spm * babystep.accum));
      #if ENABLED(BABYSTEP_DISPLAY_TOTAL)
        #if ENABLED(TFT_COLOR_UI)
          lcd_moveto(4, 3);
          lcd_put_u8str(GET_TEXT(MSG_BABYSTEP_TOTAL));
          lcd_put_wchar(':');
          lcd_moveto(10, 3);
          lcd_put_u8str(BABYSTEP_TO_STR(spm * babystep.axis_total[BS_TOTAL_IND(axis)]));
        #else
          const bool in_view = TERN1(HAS_MARLINUI_U8GLIB, PAGE_CONTAINS(LCD_PIXEL_HEIGHT - MENU_FONT_HEIGHT, LCD_PIXEL_HEIGHT - 1));
          if (in_view) {
            TERN_(HAS_MARLINUI_U8GLIB, ui.set_font(FONT_MENU));
            lcd_moveto(0, TERN(HAS_MARLINUI_U8GLIB, LCD_PIXEL_HEIGHT - MENU_FONT_DESCENT, LCD_HEIGHT - 1));
            lcd_put_u8str_P(GET_TEXT(MSG_BABYSTEP_TOTAL));
            lcd_put_wchar(':');
            lcd_put_u8str(BABYSTEP_TO_STR(spm * babystep.axis_total[BS_TOTAL_IND(axis)]));
          }
        #endif
      #endif
    }
```

give that :

image

still not perfect, cant get the ":" correctly for some reason i dont understand...

**Comment by tpruvot** [comment]
just void menu_pause_option(); to move and that build it

**Comment by thinkyhead** [comment]
Seemed like a good time to collect the common pieces in one place. More parts of those files can probably be moved into `ui_common.cpp` if more positional defines are added to the headers, or if common positioning can be used just fine where they now differ.

**Comment by tpruvot** [comment]
indeed, to avoid multiple edit on helper functions at least ;)

**Comment by tpruvot** [comment]
else, this refactor didn't broke the display... tested ok

**Comment by thinkyhead** [comment]
I don't mind merging this for wider testing at this point….

**Comment by thisiskeithb** [comment]
Give this [Anet ET5X/Anet TFT35 config](https://github.com/thisiskeithb/Marlin/tree/d5d4cf02129903444902d1b33a3acfd10d337dc0) a build. Rebased against c74f972 from ~40 mins ago.

<details><summary>Full error output:</summary>

```prolog
Compiling .pio/build/Anet_ET4_OpenBLT/src/src/lcd/tft/ui_common.cpp.o
Compiling .pio/build/Anet_ET4_OpenBLT/src/src/lcd/tft_io/tft_io.cpp.o
In file included from Marlin/src/lcd/tft/../../inc/../core/boards.h:24,
                 from Marlin/src/lcd/tft/../../inc/MarlinConfigPre.h:37,
                 from Marlin/src/lcd/tft/tft_image.cpp:23:
Marlin/src/lcd/tft/tft_image.cpp:77:75: error: 'MARLIN_LOGO_FULL_SIZE' was not declared in this scope
   77 |   TERN(SHOW_BOOTSCREEN, TERN(BOOT_MARLIN_LOGO_SMALL, MarlinLogo195x59x16, MARLIN_LOGO_FULL_SIZE), NoLogo),
      |                                                                           ^~~~~~~~~~~~~~~~~~~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:443:26: note: in definition of macro 'THIRD'
  443 | #define THIRD(a,b,c,...) c
      |                          ^
Marlin/src/lcd/tft/../../inc/../core/macros.h:192:29: note: in expansion of macro '___TERN'
  192 | #define __TERN(T,V...)      ___TERN(_CAT(_NO,T),V)  // Prepend '_NO' to get '_NOT_0' or '_NOT_1'
      |                             ^~~~Compiling .pio/build/Anet_ET4_OpenBLT/src/src/lcd/tft_io/touch_calibration.cpp.o
~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:191:29: note: in expansion of macro '__TERN'
  191 | #define _TERN(E,V...)       __TERN(_CAT(T_,E),V)    // Prepend 'T_' to get 'T_0' or 'T_1'
      |                             ^~~~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:187:29: note: in expansion of macro '_TERN'
  187 | #define TERN(O,A,B)         _TERN(_ENA_1(O),B,A)    // OPTION converted to '0' or '1'
      |                             ^~~~~
Marlin/src/lcd/tft/tft_image.cpp:77:3: note: in expansion of macro 'TERN'
   77 |   TERN(SHOW_BOOTSCREEN, TERN(BOOT_MARLIN_LOGO_SMALL, MarlinLogo195x59x16, MARLIN_LOGO_FULL_SIZE), NoLogo),
      |   ^~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:193:29: note: in expansion of macro 'THIRD'
  193 | #define ___TERN(P,V...)     THIRD(P,V)              // If first argument has a comma, A. Else B.
      |                             ^~~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:192:29: note: in expansion of macro '___TERN'
  192 | #define __TERN(T,V...)      ___TERN(_CAT(_NO,T),V)  // Prepend '_NO' to get '_NOT_0' or '_NOT_1'
      |                             ^~~~~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:191:29: note: in expansion of macro '__TERN'
  191 | #define _TERN(E,V...)       __TERN(_CAT(T_,E),V)    // Prepend 'T_' to get 'T_0' or 'T_1'
      |                             ^~~~~~
Marlin/src/lcd/tft/../../inc/../core/macros.h:187:29: note: in expansion of macro '_TERN'
  187 | #define TERN(O,A,B)         _TERN(_ENA_1(O),B,A)    // OPTION converted to '0' or '1'
      |                             ^~~~~
Marlin/src/lcd/tft/tft_image.cpp:77:25: note: in expansion of macro 'TERN'
   77 |   TERN(SHOW_BOOTSCREEN, TERN(BOOT_MARLIN_LOGO_SMALL, MarlinLogo195x59x16, MARLIN_LOGO_FULL_SIZE), NoLogo),
      |                         ^~~~
Compiling .pio/build/Anet_ET4_OpenBLT/src/src/libs/buzzer.cpp.o
Compiling .pio/build/Anet_ET4_OpenBLT/src/src/libs/crc16.cpp.o
Compiling .pio/build/Anet_ET4_OpenBLT/src/src/libs/nozzle.cpp.o
Compiling .pio/build/Anet_ET4_OpenBLT/src/src/libs/numtostr.cpp.o
*** [.pio/build/Anet_ET4_OpenBLT/src/src/lcd/tft/tft_image.cpp.o] Error 1
```
</details>

https://github.com/MarlinFirmware/Marlin/pull/20861 appears to be the bug source according to a little `git blame`.

I have `BOOT_MARLIN_LOGO_SMALL` disabled, but enabling it causes the build to succeed.

**Pull request merged**